python/lsst/ts/genericcamera/genericcameracsc.py

```diff
@@ -705,7 +626,164 @@ async def run_auto_exposure_loop(self, min_exp_time, max_exp_time, configuration
         self.isAutoExposure = False
         self.log.info("autoExposure_loop - End")
 
-    async def establish_exposure_background(self, exposure):
+    async def run_auto_exposure(self, min_exp_time, max_exp_time, configuration):
+        """Take auto exposures until `runAutoExposureTask` is False.
+
+        Parameters
+        ----------
+        min_exp_time: `float`
+            The minimum exposure time to use.
+        max_exp_time: `float`
+            The maximum exposure time to use.
+        configuration: `dict`
+            A dict containing additional configuration parameters.
+        """
+        while self.runAutoExposureTask:
```

Comment:
I think we are converging but I still have some suggestions about this method. One problem here is that you are running a full "determine" loop at every iteration. But the exposure time will most likely change very slowly over time, except maybe during twilight. Here's what I think this method should look like:

```
async def run_auto_exposure(self, ...):
    """
    ...
    """
    self.log.debug("Determining initial auto exposure time.")
    exposure_time_auto_current = await self.determine_auto_exposure_time(...)

   self.log.debug(f"Initial auto exposure time: {exposure_time_auto_current}s [{min_exp_time}:{max_exp_time}]")

   while self.runAutoExposureTask:

        ...

        exposure = await self.take_image(
                 configuration["shutter"],
                 configuration["images_in_sequence"],
                 configuration["image_index"],
                 exposure_time_auto_current,
                 timestamp,
                 image_name,
             )

        ...

        background_level = self.establish_exposure_background(exposure)

        exposure_time_auto_new = self.adjust_exposure_time(
                exposure_time_auto_current,
                background_level, 
                ....
             )
        self.log.debug(f""Auto exposure time adjusted: {exposure_time_auto_current}s -> {exposure_time_auto_new}s.)
        exposure_time_auto_current = exposure_time_auto_new
        ....
```

Reply:
Thanks very much for the extensive explanation. I reread your original comment and that indeed is what you wrote before but I failed to understand that. I have reviewed the code and it looks like this was largely in place already. I updated the code to follow this pattern.

Reply:
It's getting close, but not quite there yet. You are still calling `self.get_auto_exposure_and_exposure_time` at every iteration loop. 

I think what you need to do is call `self.get_auto_exposure_and_exposure_time` once, before entering the loop. Then, once inside the loop you will take only 1 exposure (with the exposure time estimated by `self.get_auto_exposure_and_exposure_time`) and perform 1 adjustment of the exposure time based on that image.